airflow/datasets/__init__.py

```diff
@@ -73,6 +73,12 @@ def __eq__(self, other):
     def __hash__(self):
         return hash(self.uri)
 
+    def __or__(self, other: Dataset):
```

Comment:
```suggestion
    def __or__(self, other: Dataset | DatasetAny | DatasetAll) -> DatasetAny:
```

Can't other be anything?

Reply:
Changed it to `BaseDatasetEventInput`